kythe/go/platform/tools/kzip/viewcmd/viewcmd.go

```diff
@@ -0,0 +1,169 @@
+/*
+ * Copyright 2019 The Kythe Authors. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// Package viewcmd displays the contents of compilation units stored in .kzip files.
+package viewcmd // import "kythe.io/kythe/go/platform/tools/kzip/viewcmd"
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/golang/protobuf/jsonpb"
+	"github.com/golang/protobuf/proto"
+	"github.com/google/subcommands"
+	"golang.org/x/sync/errgroup"
+	"golang.org/x/sync/semaphore"
+	"kythe.io/kythe/go/platform/kzip"
+	"kythe.io/kythe/go/util/cmdutil"
+)
+
+type cmd struct {
+	cmdutil.Info
+	extractDir string
+}
+
+// New returns an implementation of the "view" subcommand.
+func New() subcommands.Command {
+	const usage = `Usage: view [options] <file-path>...
+
+Print or extract compilation units stored in .kzip files.  By default, the
+compilation unit is printed to stdout in JSON format.
+
+With -extract the compilation record and the full contents of all the required
+input files are extracted into the named directory, preserving the path
+structure specified in the compilation record.`
+
+	return &cmd{
+		Info: cmdutil.NewInfo("view", "view compilation units stored in .kzip files", usage),
+	}
+}
+
+// SetFlags implements part of subcommands.Command.
+func (c *cmd) SetFlags(fs *flag.FlagSet) {
+	fs.StringVar(&c.extractDir, "extract", "", "Extract files to this directory")
+}
+
+// Execute implements part of subcommands.Command.
+func (c *cmd) Execute(ctx context.Context, fs *flag.FlagSet, args ...interface{}) subcommands.ExitStatus {
+	hasErrors := false
+	for _, path := range fs.Args() {
+		ext := filepath.Ext(path)
+		base := filepath.Base(strings.TrimSuffix(path, ext))
+
+		f, err := os.Open(path)
+		if err != nil {
+			log.Printf("Error opening .kzip file: %v", err)
+			hasErrors = true
+			continue
+		}
+		defer f.Close()
+
+		err = kzip.Scan(f, func(r *kzip.Reader, unit *kzip.Unit) error {
+			if err := c.writeUnit(base+"-"+unit.Digest, unit.Proto); err != nil {
+				return fmt.Errorf("writing unit: %v", err)
+			} else if fd, err := c.kzipFiles(r, unit); err != nil {
```

Comment:
If we're deprecating `viewindex` we should allow for its behavior of only printing the compilation unit without its required inputs.

Reply:
This is the behavior if `--extract` is unset. `writeFiles()` and `kzipFiles()` are both no-ops if an extract directory is not given.

If it'd be clearer, I can move the `extractDir` check up here in `Execute()`.